jlm/util/Statistics.cpp

```diff
@@ -1,15 +1,171 @@
+    { Statistics::Id::AgnosticMemoryNodeProvisioning, "AgnosticMemoryNodeProvider" },
+    { Statistics::Id::AndersenAnalysis, "AndersenAnalysis" },
+    { Statistics::Id::Annotation, "Annotation" },
+    { Statistics::Id::CommonNodeElimination, "CNE" },
+    { Statistics::Id::ControlFlowRecovery, "ControlFlowRestructuring" },
+    { Statistics::Id::DataNodeToDelta, "DataNodeToDeltaStatistics" },
+    { Statistics::Id::DeadNodeElimination, "DeadNodeElimination" },
+    { Statistics::Id::FunctionInlining, "ILN" },
+    { Statistics::Id::JlmToRvsdgConversion, "ControlFlowGraphToLambda" },
+    { Statistics::Id::LoopUnrolling, "UNROLL" },
+    { Statistics::Id::InvariantValueRedirection, "InvariantValueRedirection" },
+    { Statistics::Id::MemoryStateEncoder, "MemoryStateEncoder" },
+    { Statistics::Id::PullNodes, "PULL" },
+    { Statistics::Id::PushNodes, "PUSH" },
+    { Statistics::Id::ReduceNodes, "RED" },
+    { Statistics::Id::RegionAwareMemoryNodeProvisioning, "RegionAwareMemoryNodeProvision" },
+    { Statistics::Id::RvsdgConstruction, "InterProceduralGraphToRvsdg" },
+    { Statistics::Id::RvsdgDestruction, "RVSDGDESTRUCTION" },
+    { Statistics::Id::RvsdgOptimization, "RVSDGOPTIMIZATION" },
+    { Statistics::Id::SteensgaardAnalysis, "SteensgaardAnalysis" },
+    { Statistics::Id::ThetaGammaInversion, "IVT" }
+  };
+  // Make sure every Statistic is mentioned in the mapping
+  auto lastIdx = static_cast<size_t>(Statistics::Id::LastEnumValue);
+  auto firstIdx = static_cast<size_t>(Statistics::Id::FirstEnumValue);
+  JLM_ASSERT(mapping.Size() == lastIdx - firstIdx - 1);
+  return mapping;
+}
+
 Statistics::~Statistics() = default;
 
+std::string_view
+Statistics::GetName() const
+{
+  return GetStatisticsIdNames().LookupKey(StatisticsId_);
+}
+
+const util::filepath &
+Statistics::GetSourceFile() const
+{
+  return SourceFile_;
+}
+
+std::string
+Statistics::ToString() const
+{
+  return util::strfmt(GetName(), " ", GetSourceFile().to_str(), " ", Serialize());
+}
+
+std::string
+Statistics::Serialize() const
+{
+  std::ostringstream ss;
+  for (const auto & [mName, measurement] : Measurements_)
+  {
+    if (ss.tellp() != 0)
+      ss << " ";
+
+    ss << mName << ":";
+    std::visit(
+        [&](const auto & value)
+        {
+          ss << value;
+        },
+        measurement);
+  }
+  for (const auto & [mName, timer] : Timers_)
+  {
+    if (ss.tellp() != 0)
+      ss << " ";
+
+    ss << mName << "[ns]:" << timer.ns();
+  }
+
+  return ss.str();
+}
+
+bool
+Statistics::HasMeasurement(const std::string & name) const noexcept
+{
+  for (const auto & [mName, _] : Measurements_)
+    if (mName == name)
+      return true;
+  return false;
+}
+
+Statistics::Measurement &
+Statistics::GetMeasurement(const std::string & name)
+{
+  for (auto & [mName, measurement] : Measurements_)
+    if (mName == name)
+      return measurement;
+  throw util::error(util::strfmt("Unknown measurement: ", name));
```

Comment:
I would almost like to replace this with `JLM_ASSERT()`. Nobody catches it anyway, and it is basically a bug to ask for a measurement that is not present.